### PR TITLE
WIP: PARQUET-1095/PARQUET-982: [C++] Read and write Arrow decimal values

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

**Comment by wesm** [comment]
I believe that if we use FIXED_LEN_BYTE_ARRAY, that we may need to pack the data into the smallest byte width that will accommodate the decimal precision. We should double check about whether Spark or other Java consumers are able to decode this data

File: src/parquet/arrow/reader.cc [hunk 1]
```diff
@@ -1319,29 +1336,21 @@ Status PrimitiveImpl::NextBatch(int batch_size, std::shared_ptr<Array>* out) {
...
-      break;
-    }
+    case ::arrow::Type::FIXED_SIZE_BINARY:
+      return ReadFLBABatch<::arrow::FixedSizeBinaryType>(batch_size, type, out);
+    case ::arrow::Type::DECIMAL:
+      return ReadFLBABatch<::arrow::DecimalType>(batch_size, type, out);
     case ::arrow::Type::TIMESTAMP: {
-      ::arrow::TimestampType* timestamp_type =
-          static_cast<::arrow::TimestampType*>(field_->type().get());
+      const auto& timestamp_type = std::static_pointer_cast<::arrow::TimestampType>(type);
```
Comment (xhochy) [inline src/parquet/arrow/reader.cc]: As you already touch this line, change this to `static_cast<const ::arrow::TimestampType&>`. This should save us from some unnecessary reference counting.

File: src/parquet/arrow/writer.cc [hunk 2]
```diff
@@ -679,7 +678,7 @@ Status FileWriter::Impl::TypedWriteBatch<BooleanType, ::arrow::BooleanType>(
     const int16_t* def_levels, const int16_t* rep_levels) {
   RETURN_NOT_OK(data_buffer_.Resize(array->length()));
   auto data = static_cast<const BooleanArray*>(array.get());
-  auto data_ptr = reinterpret_cast<const uint8_t*>(data->values()->data());
+  auto data_ptr = data->values()->data();
```
Comment (xhochy) [inline src/parquet/arrow/writer.cc]: Google styleguide nitpick: Only use `auto` where the type of the variable is already mentioned in the line.

**Comment by wesm** [comment]
I have a somewhat invasive refactor of parquet/arrow/reader.cc ongoing for PARQUET-1100, I will try to get it to ship shape in the next couple days

**Comment by cpcloud** [comment]
I'm going to wait for PARQUET-1100 before continuing on this patch.

**Pull request closed without merging**